### PR TITLE
Adds functionality to find_MAP

### PR DESCRIPTION
**Pull request by bjedwards**
Allows for the use of minimization functions that do not require a gradient. Also adds basic tests for find_MAP. Addresses concerns in issue #396.

### CODE REVIEW

**Comment by bjedwards** [comment]
Whoops, forgot about the deprecation of `func_code` in python 3, will fix.

File: pymc/tuning/starting.py [hunk 1]
```diff
@@ -50,9 +50,14 @@ def logp_o(point):
 
     def grad_logp_o(point):
         return nan_to_num(-dlogp(point))
+    
+    # Check to see if minimization function actually uses the gradient
+    if 'fprime' in fmin.__code__.co_varnames:
```
Comment (jsalvatier) [inline pymc/tuning/starting.py]: I think it would be better to use: `'fprime' in getargspec(fmin).args`

File: pymc/tests/test_tuning.py [hunk 2]
```diff
@@ -16,3 +17,25 @@ def test_guess_scaling():
...
+        mu = Uniform('mu',-1,1)
+        sigma = Uniform('sigma',.5,1.5)
+        y = Normal('y',mu=mu,tau=sigma**-2,observed=data)
+        
+        # Test gradient minimization
+        map_est1 = starting.find_MAP()
+        # Test non-gradient minimization
+        map_est2 = starting.find_MAP(fmin=starting.optimize.fmin_powell)
+    
+    assert (-tol<=map_est1['mu']<=tol)
```
Comment (jsalvatier) [inline pymc/tests/test_tuning.py]: Might use `close_to` instead of these asserts

```
import close_to from checks
...
close_to(map_est1['mu'], 0, tol)
```

**Comment by jsalvatier** [comment]
Very clean and adds some useful functionality and tests. Thanks for this! Only some minor changes if you have time, then good to merge.

**Comment by bjedwards** [comment]
Updated, based on your suggestion. I knew something like `close_to` had to exist somewhere, I just didn't look hard enough.

File: pymc/tests/test_tuning.py [hunk 3]
```diff
@@ -16,3 +18,26 @@ def test_guess_scaling():
...
+        mu = Uniform('mu',-1,1)
+        sigma = Uniform('sigma',.5,1.5)
+        y = Normal('y',mu=mu,tau=sigma**-2,observed=data)
+        
+        # Test gradient minimization
+        map_est1 = starting.find_MAP()
+        # Test non-gradient minimization
+        map_est2 = starting.find_MAP(fmin=starting.optimize.fmin_powell)
+    
+    close_to(map_est1['mu'],0,tol)
```
Comment (twiecki) [inline pymc/tests/test_tuning.py]: We are mildly enforcing pep8 which says white-spaces after commas.

See flake8  to spot these or run autopep8 to fix them automatically.

**Comment by bjedwards** [comment]
I think this should be ready to go, there are a couple of lines in `tuning\starting.py`, reporting an error message that are over 80 characters. I can reformat if like.

**Comment by twiecki** [comment]
Awesome, thanks!

**Comment by jsalvatier** [comment]
:D

**Pull request merged**